api/apps/geoprocessing/src/modules/shapefiles/shapefiles.service.ts

```diff
@@ -1,33 +1,80 @@
-import { Injectable, Logger } from '@nestjs/common';
+import { BadRequestException, Injectable, Logger } from '@nestjs/common';
+import { readdir } from 'fs/promises';
+import { FeatureCollection } from 'geojson';
 import { FileService } from '../files/files.service';
+
 const mapshaper = require('mapshaper');
 
 @Injectable()
 export class ShapefileService {
   private readonly logger: Logger = new Logger(ShapefileService.name);
+  private readonly minRequiredFiles = ['prj', 'dbf', 'shx', 'shp'];
   constructor(private readonly fileService: FileService) {}
 
   private async shapeFileToGeoJson(fileInfo: Express.Multer.File) {
+    if (
+      !(await this.areRequiredShapefileFilesInFolder(
+        fileInfo.path.replace('.zip', ''),
+      ))
+    ) {
+      throw new Error(
+        'Shapefile data is not usable: one or more required files are missing.',
+      );
+    }
     const outputKey = `shapefile-${new Date().getTime()}.geojson`;
 
     const _geoJson = await mapshaper.applyCommands(
-      `-i ${fileInfo.path.replace('.zip', '')}/*.shp -info -o ${outputKey}`,
+      `-i no-topology snap ${fileInfo.path.replace(
+        '.zip',
+        '',
+      )}/*.shp -proj EPSG:4326 -clean rewind -info -o ${outputKey}`,
     );
 
     return JSON.parse(_geoJson[outputKey].toString('utf-8'));
   }
 
+  async areRequiredShapefileFilesInFolder(path: string): Promise<boolean> {
+    const filesInPath = await readdir(path);
+    const extensions = filesInPath.map((file) => file.split('.').pop());
+    return this.minRequiredFiles.every((ext: string) =>
+      extensions.includes(ext),
+    );
+  }
+
+  isValidGeoJson(geoJson: FeatureCollection): boolean {
```

Comment:
1. if you are checking GeoJson, put relevant type instead concrete one. 
2. It does not actually check if GeoJson is valid - it checks if given Geo is a Feature Collection and contains only relevant types (i.e. `isGeoJsonTypeSupported`)

Reply:
Changes: ad322a92ea39795a68cd1db0d2cdeff3d1dec48d